nuxeo-core-binarymanager-s3/src/main/java/org/nuxeo/ecm/core/storage/sql/CloudFrontBinaryManager.java

```diff
@@ -0,0 +1,156 @@
+    protected String distributionDomain;
+
+    protected Protocol protocol;
+
+    protected PrivateKey privKey;
+
+    protected String privKeyId;
+
+    @Override
+    protected void setupCloudClient() throws IOException {
+        super.setupCloudClient();
+
+        protocol = Protocol.valueOf(getProperty(PROTOCOL_PROPERTY, "https"));
+        distributionDomain = getProperty(DISTRIB_DOMAIN_PROPERTY);
+
+        try {
+            String privateKeyPath = getProperty(PRIVATE_KEY_PROPERTY);
+            privKey = loadPrivateKey(privateKeyPath);
+            privKeyId = getProperty(PRIVATE_KEY_ID_PROPERTY);
+        } catch (InvalidKeySpecException e) {
+            throw new IOException(e);
+        }
+    }
+
+    @Override
+    protected URI getRemoteUri(String digest, ManagedBlob blob, HttpServletRequest servletRequest) throws IOException {
+        Date expiration = new Date();
+        expiration.setTime(expiration.getTime() + directDownloadExpire * 1000);
+
+        try {
+            URIBuilder uriBuilder = new URIBuilder(buildResourcePath(bucketNamePrefix + digest));
+            if (blob != null) {
+                uriBuilder.addParameter("response-content-type", getContentTypeHeader(blob));
+                uriBuilder.addParameter("response-content-disposition",
+                        getContentDispositionHeader(blob, servletRequest));
+            }
+
+            String uri;
+            if (privKey != null) {
+                uri = CloudFrontUrlSigner.getSignedURLWithCannedPolicy(uriBuilder.build().toString(), privKeyId,
+                        privKey, expiration);
+            } else {
+                uri = uriBuilder.build().toString();
+            }
+
+            return new URI(uri);
+        } catch (URISyntaxException e) {
+            throw new IOException(e);
+        }
+    }
+
+    protected String buildResourcePath(String s3ObjectKey) {
+        return protocol != CloudFrontUrlSigner.Protocol.http && protocol != CloudFrontUrlSigner.Protocol.https
+                ? s3ObjectKey : protocol + "://" + distributionDomain + "/" + s3ObjectKey;
+    }
+
+    /**
+     * Originally from CloudFrontUrlSigner.loadPrivateKey()
+     */
+    static PrivateKey loadPrivateKey(String privateKeyPath) throws InvalidKeySpecException, IOException {
+        if (privateKeyPath == null) {
+            return null;
+        }
+
+        File privateKeyFile = new File(privateKeyPath);
+
+        FileInputStream is;
+        PrivateKey var2;
+        if (privateKeyFile.getAbsolutePath().toLowerCase().endsWith(".pem")) {
+            is = new FileInputStream(privateKeyFile);
+
+            try {
+                var2 = PEM.readPrivateKey(is);
+            } finally {
+                try {
+                    is.close();
+                } catch (IOException var19) {
+                    ;
+                }
+
+            }
+
+            return var2;
+        } else if (privateKeyFile.getAbsolutePath().toLowerCase().endsWith(".der")) {
```

Comment:
Factor the `new FileInputStream` which is present in the two branches of the `if`/`else`